File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Action/Common/GainBlockAction.ts
import Async from "../../../../../Core/Async";
import { AttackEffect } from "../../DataDefine/AttackEffect";
import DungeonManager from "../../DungeonManager";
import { FlashAtkImgEffect } from "../../Effect/FlashAtkImgEffect";
import AbstractCreature from "../../Unit/AbstractCreature";
import AbstractGameAction, { ActionType } from "../AbstractGameAction";

export default class GainBlockAction extends AbstractGameAction
{
    private Target:AbstractCreature;
    private Source:AbstractCreature;
    private Amount:number;
    Type = ActionType.BLOCK;
    public constructor(target : AbstractCreature, amount : number, source: AbstractCreature = null) {
        super();
        this.Target = target;
        if(!source)
            this.Source = source;
        else
            this.Source = target;
        this.Amount = amount;
    }
    
    async Update(){
        if(this.Target && !this.Target.IsDying && !this.Target.IsDead){
            var hitbox = this.Target.GetHitBox();
            DungeonManager.EffectManager.Play(new FlashAtkImgEffect(hitbox.CX,hitbox.CY,AttackEffect.SHIELD))
            this.Target.Block += this.Amount;
            return true;
        }
        await Async.Delay(250);
        return true;
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Action/Combat/GainBlockAction.cs
using Cysharp.Threading.Tasks;
using ModTheSpire.Dungeon;
using ModTheSpire.Effect.Combat;
using ModTheSpire.Event;
using ModTheSpire.Unit;

namespace ModTheSpire.Effect.Other
{
    public class GainBlockAction : IGameAction
    {
        private readonly AbstractCreature mTarget;
        private readonly int mAmount;
        private readonly AbstractCreature mSource;
        public float Duration { get; set; }
        public ActionType Type { get; } = ActionType.BLOCK;
        public void Init()
        {
        }

        public GainBlockAction(AbstractCreature target, int amount, AbstractCreature source = null)
        {
            mTarget = target;
            mAmount = amount;
            mSource = source;
        }

        public async UniTask<bool> Update()
        {
            if(this.mTarget != null && !this.mTarget.IsDying && !this.mTarget.IsDead){
                var hitbox = this.mTarget.DisplayObject.GetHitBox();
                DungeonManager.EffectManager.Play(new FlashAtkImgEffect(hitbox.center.x, hitbox.center.y, AttackEffect.SHIELD));
                this.mTarget.Block += this.mAmount;
                await UniTask.Delay(250);
                return true;
            }
            return true;
        }
    }
}